### DISCUSSION TITLE
Handling field names that conflict with serializer attributes?

### DISCUSSION THREAD

**Discussion by Kangaroux** [encode/django-rest-framework#8969] [Question & Answer]
I have a data structure I need to validate which has a field called `label`. However, serializers have an attribute on them called `label`. Trying to set the field is overwriting the attribute which mypy (correctly) complains about:

```python
class FooSerializer(serializers.Serializer):
    # Error: Incompatible types in assignment (expression has type "CharField", 
    # base class "Field" defined the type as "Optional[str]")  [assignment] mypy
    label = serializers.CharField(max_length=10)
```

I can rename the field and change the `source` so that it will serialize the object correctly:

```python
label_ = serializers.CharField(max_length=10, source="label")
```

However, the `source` argument only affects serializing, not deserializing:

```python
>>> s = FooSerializer(data={"label": "test"})
>>> s.is_valid()
False
>>> s.errors
{'label_': [ErrorDetail(string='This field is required.', code='required')]}
```

Is it possible to do this without hacking together something in an override?

**Comment by TauPan** [comment]
Hi!

This is an old question, but I found this while looking for answers to my problem, so I'll add this for posterity. (Because of [Wisdom of the Ancients](https://xkcd.com/979/))

I had a similar problem in my code recently. My model has a `JSONField` called `data` which is also a property on the serializer class and object. (The distinction between object and class **is** important in this case.)

This works fine in practice, since DRF moves all Fields in the object to `_declared_fields` so the serializer **object** still has `data` as a property, while the declared field lives in `_declared_fields` on the object. Of course neither mypy nor pyright can grok this.

What I did was to declare the field as a field, which satisfies both mypy and pyright:

`data: serializers.Field = JSONField(required=False)` 

In the OP's case the following should help:

~~~python
class FooSerializer(serializers.Serializer):
    # All serializer Fields are moved into _declared_fields when the
    # metaclass creates the Serializer object, but the type checker
    # will think that `data` is a property, because it is defined as
    # such, so we have to assure it, that, only in this subclass, it's
    # a field.
    label: serializers.Field = serializers.CharField(max_length=10)
~~~

Cheers!
